src/storage.rs

```diff
@@ -419,200 +450,324 @@ impl Cache {
+                    })
+                    .await?
+                }
+            })
+            .await?;
+        Ok(path.to_owned())
     }
 
-    pub fn fetch_and_diffstat_all(
-        &mut self,
-        network: Option<Arc<Network>>,
-        package: PackageStr,
-        diffs: &SortedSet<Delta>,
-    ) -> Result<DiffRecommendation, VetError> {
-        let _span = trace_span!("diffstat-all").entered();
-        // If there's no registry path setup, assume we're in tests and mocking.
-        let mut all_versions = SortedSet::new();
-
-        for delta in diffs {
-            let is_cached = self
+    #[tracing::instrument(skip(self, network))]
+    pub async fn fetch_and_diffstat_package(
+        &self,
+        network: &Option<Arc<Network>>,
+        package: PackageStr<'_>,
+        delta: &Delta,
+    ) -> Result<DiffStat, VetError> {
+        // Use a once_cell to cache the result here if we try to fetch the same
+        // package multiple times from a single execution.
+        let once_cell = {
+            let mut guard = self.state.lock().unwrap();
+
+            // Check if the value has already been cached.
+            if let Some(cached) = guard
                 .diff_cache
                 .get(package)
                 .and_then(|cache| cache.get(delta))
-                .is_some();
-            if !is_cached {
-                all_versions.insert(&delta.from);
-                all_versions.insert(&delta.to);
+                .cloned()
+            {
+                return Ok(cached);
             }
-        }
 
-        let mut best_rec: Option<DiffRecommendation> = None;
-        let to_fetch = all_versions
-            .iter()
-            .map(|v| (package, *v))
-            .collect::<Vec<_>>();
-        let fetches = self.fetch_packages(network, &to_fetch)?;
+            if self.root.is_none() {
+                // If we don't have a root, assume we want mocked results
+                // ERRORS: this warning really rides the line, I'm not sure if the user can/should care
+                warn!("Missing root, assuming we're in tests and mocking");
+
+                let from_len = delta.from.major * delta.from.major;
+                let to_len: u64 = delta.to.major * delta.to.major;
+                let diff = to_len as i64 - from_len as i64;
+                let count = diff.unsigned_abs();
+                let raw = if diff < 0 {
+                    format!("-{}", count)
+                } else {
+                    format!("+{}", count)
+                };
+                return Ok(DiffStat { raw, count });
+            }
 
-        for delta in diffs {
-            let cached = self
-                .diff_cache
-                .get(package)
-                .and_then(|cache| cache.get(delta))
-                .cloned();
+            guard
+                .diffed
+                .entry((package.to_owned(), delta.clone()))
+                .or_default()
+                .clone()
```

Comment:
Wait I only just now realized you're taking a Mutex to the state and then cloning the interior once_cell out... ohhh it's an Arc ok ok. This might merit a comment at the two places you do this because it seems super wrong in a vacuum.

Reply:
Yeah, I think I'm going to refactor this into a helper type like `CacheMap<K, V>` at some point, which internally is a `Mutex<HashMap<K, Arc<OnceCell<V>>>>`.